core/src/main/java/org/elasticsearch/cluster/metadata/SnapshotId.java

```diff
@@ -19,37 +19,78 @@
 
 package org.elasticsearch.cluster.metadata;
 
+import org.elasticsearch.common.UUIDs;
 import org.elasticsearch.common.io.stream.StreamInput;
 import org.elasticsearch.common.io.stream.StreamOutput;
-import org.elasticsearch.common.io.stream.Streamable;
+import org.elasticsearch.common.io.stream.Writeable;
+import org.elasticsearch.snapshots.Snapshot;
 
 import java.io.IOException;
+import java.util.Objects;
 
 /**
- * Snapshot ID - repository name + snapshot name
+ * SnapshotId - snapshot (repository name + snapshot name) + snapshot UUID
  */
-public class SnapshotId implements Streamable {
+public class SnapshotId implements Writeable {
 
-    private String repository;
+    public static final String UNASSIGNED_UUID = ""; // empty string so its compatible with old blob naming
 
-    private String snapshot;
+    private final Snapshot snapshot;
+    private final String uuid;
 
     // Caching hash code
-    private int hashCode;
+    private final int hashCode;
 
-    private SnapshotId() {
+    /**
+     * Constructs a new snapshot
+     *
+     * @param snapshot snapshot description
+     * @param uuid     snapshot uuid
+     */
+    private SnapshotId(final Snapshot snapshot, final String uuid) {
+        this.snapshot = Objects.requireNonNull(snapshot);
+        this.uuid = Objects.requireNonNull(uuid);
+        this.hashCode = computeHashCode();
     }
 
     /**
-     * Constructs new snapshot id
+     * Constructs a new snapshot from a input stream
      *
-     * @param repository repository name
-     * @param snapshot   snapshot name
+     * @param in  input stream
      */
-    public SnapshotId(String repository, String snapshot) {
-        this.repository = repository;
-        this.snapshot = snapshot;
-        this.hashCode = computeHashCode();
+    public SnapshotId(final StreamInput in) throws IOException {
+        snapshot = new Snapshot(in);
+        uuid = in.readString();
+        hashCode = computeHashCode();
+    }
+
+    /**
+     * Create a new SnapshotId with a new UUID.  This should be used when getting a
+     * SnapshotId for a new snapshot.
+     *
+     * @param snapshot snapshot
+     * @return snapshot id
+     */
+    public static SnapshotId createNew(final Snapshot snapshot) {
+        return new SnapshotId(snapshot, UUIDs.randomBase64UUID());
+    }
+
+    /**
+     * Get a SnapshotId from the given components.
+     *
+     * @param snapshot   the snapshot description
+     * @param uuid       the snapshot uuid
+     * @return snapshot id
+     */
+    public static SnapshotId get(final Snapshot snapshot, final String uuid) {
+        return new SnapshotId(snapshot, uuid);
```

Comment:
As this is just a public wrapper for new Snapshot(...), I prefer that we make the constructor public and get rid of this method.